File: TicTacToe.scala
import scala.collection.mutable.ArrayBuffer

case class TicTacToe(len: Int) {
  private val board = ArrayBuffer.fill[Option[Char]](len,len)(None)

  def move(player: Char, row: Int, col: Int): Unit = ???

  def winner: Option[Char] =
    (rowWinner ++ colWinner ++ diaWinner)
      .filter(_.isDefined)
      .headOption
      .flatten

  private def rowWinner = board.map(winnerOf)

  private def colWinner =
    (0 until len)
      .map(i => board.map(_(i)))
      .map(winnerOf)

  private def diaWinner =
    Seq(
      (0 until len).map(i => board(i)(len-1-i)),
      (0 until len).map(i => board(i)(i))
    ).map(winnerOf)

  private def winnerOf(a: Seq[Option[Char]]): Option[Char] = {
    if(a(0).isDefined && a.distinct.size == 1) a(0)
    else None
  }

  def printBoard: Unit = ???
}

File: TicTacToe.java
public class TicTacToe {
  private int len;
  private int[][] board;

  public TicTacToe(int size) {
    len = size;
    board = new int[len][len];
  }

  public Character winner() {
    Character winner = rowWinner();
    if (winner != null) return winner;
    winner = colWinner();
    if (winner != null) return winner;
    return diagWinner();
  }

  private Character rowWinner() {
    Character atZero = null;
    for (int row = 0; row < len; row++) {
      atZero = board[row][0];
      for (int col = 1; col < len; col++)
        if (!atZero.equals(board[row][col])) break;
        if (col == len - 1) return atZero;
    }
    return null;
  }

  private Character colWinner() {
    Character atZero = null;
    for (int col = 0; col < len; col++) {
      atZero = board[0][col];
      for (int row = 1; row < len; row++)
        if (!atZero.equals(board[row][col])) break;
        if (row == len - 1) return atZero;
    }
    return null;
  }

  private Character diagWinner() {
    Character atZero = board[0][0];
    for (int row = 1, col = 1; row < len; row++, col++)
      if (!atZero.equals(board[row][col])) break;
      if (row == len - 1) return atZero;
    atZero = board[len-1][0];
    for (int row = len-2, col = 1, col < len; row--, col++)
      if (!atZero.equals(board[row][col])) break;
      if (col == len - 1) return atZero;
    return null;
  }
}
